### PR TITLE
Migration Bug Fixes

### PR DESCRIPTION
**Pull request by zacheliason**

### CODE REVIEW

File: front_end/content.py [hunk 1]
```diff
@@ -2006,10 +2009,14 @@ def rebuild_exercises(self, assignment_title):
 
             text_output, image_output, tests = exec_code(self.__settings_dict, exercise_details["answer_code"], exercise_basics, exercise_details)
 
+            tests_dict = []
+            for i in range(len(tests)):
+                tests_dict.append({**tests[i], **exercise_details["tests"][i]})
```
Comment (srp33) [inline front_end/content.py:2014]: For consistency with code we're using elsewhere, can we change this to use syntax more like the following:

```
for i in range(len(tests)):
    tests_dict[tests[i]] = exercise_details["tests][i]
```

File: front_end/webserver.py [hunk 2]
```diff
@@ -1106,9 +1106,11 @@ def post(self, course, assignment, exercise):
                                     tests_dict = []
                                     for i in range(len(tests)):
                                         tests_dict.append({**tests[i], **exercise_details["tests"][i]})
```
Comment (srp33) [inline front_end/webserver.py:1108]: Please modify this code, too, to be consistent with dictionary syntax we are using elsewhere.

**Review by srp33 — CHANGES_REQUESTED** [review]

File: front_end/content.py (hunk 1 above)
Comment (zacheliason) [reply front_end/content.py:2016]: I'm unfamiliar with this syntax, you'll have to show me how this works in our meeting.

File: front_end/content.py [hunk 3]
```diff
@@ -76,10 +76,9 @@ def create_database_tables(self):
                           name text,
                           given_name text,
                           family_name text,
```
Comment (zacheliason) [inline front_end/content.py:78]: The following changes are just reverting the function 'create_database_tables' to how it was before; in past commits I didn't realize that I shouldn't touch this function and in changing the way migration scripts run just now I stumbled into a few problems. Everything works again though!

File: front_end/content.py [hunk 4]
```diff
@@ -1343,7 +1341,7 @@ def get_submission_info(self, course, assignment, exercise, user, submission):
 
         row = self.fetchone(sql, (int(course), int(assignment), int(exercise), user, int(submission),))
 
-        test_sql = '''SELECT test_id, text_output, image_output
```
Comment (zacheliason) [inline front_end/content.py:1346]: renamed for clarity, since the ```tests``` table already has a ```test_id```

File: front_end/content.py [hunk 5]
```diff
@@ -1988,49 +1983,78 @@ def remove_export_paths(self, zip_file_path, tmp_dir_path):
...
             assignment = row["assignment_id"]
             exercise = row["exercise_id"]
-            print(f"Rebuilding course {course}, assignment {assignment}, exercise {exercise}")
 
             exercise_basics = self.get_exercise_basics(course, assignment, exercise)
             exercise_details = self.get_exercise_details(course, assignment, exercise)
 
             text_output, image_output, tests = exec_code(self.__settings_dict, exercise_details["answer_code"], exercise_basics, exercise_details)
 
+            exercise_details["tests"] = [{k: v for k, v in d.items() if k != 'text_output' and k != "image_output"} for d in exercise_details["tests"]]
```
Comment (zacheliason) [inline front_end/content.py:2010]: These keys have to be removed before the dict is squashed with test_outputs or else all outputs will default to ```None```

File: front_end/html/edit_exercise.html [hunk 6]
```diff
@@ -848,15 +848,13 @@ <h2>Create exercise</h2>
...
 
-            {% if len(old_tests) > 0 %}
-                var new_tests = {}
+            var new_tests = {}
 
-                {% for test in exercise_details["tests"] %}
-                    new_tests[`{{int(test['test']) - 1}}`] = {'image_output': `{{test['image_output']}}`, 'text_output': `{{test['text_output']}}` }
-                {% end %}
+            {% for i in range(len(exercise_details["tests"])) %}
+                new_tests[`{{i}}`] = {'image_output': `{{exercise_details["tests"][i]['image_output']}}`, 'text_output': `{{exercise_details["tests"][i]['text_output']}}` }
```
Comment (zacheliason) [inline front_end/html/edit_exercise.html:854]: This reconfiguration addresses a miscellaneous bug I found that arises in a really specific case.

File: front_end/html/exercise.html [hunk 7]
```diff
@@ -468,18 +468,20 @@ <h6>Data files:</h6>
...
-                        <i class="far fa-question-circle"></i>
-                    </label>
-                </div>
-                {% if exercise_details["output_type"] == "txt" %}
-                    <p><pre>{{ exercise_details["expected_text_output"] }}</pre></p>
-                {% elif exercise_details["expected_image_output"] != "" %}
-                    <p><figure class="img"><img src="data:image/jpg;base64,{{ exercise_details['expected_image_output'] }}" width="100%"/></figure></p>
-                {% else %}
-                    <p><pre>[Expected solution does not produce an image.]</pre></p>
+                {% if exercise_details["expected_text_output"] != "" or exercise_details["expected_image_output"] != "" %}
```
Comment (zacheliason) [inline front_end/html/exercise.html:471]: This hides ```Expected output``` if none is expected

File: front_end/html/exercise_shared.html [hunk 8]
```diff
@@ -192,7 +192,13 @@ <h6>Copy exercise within the same assignment:</h6>
                 }
 
                 function build_tests_output(json_data) {
-                    var test_output = "<h6>Tests outputs:</h6>\n\n";
+                    var expected_tests = document.getElementById("expected_tests");
```
Comment (zacheliason) [inline front_end/html/exercise_shared.html:195]: Addresses a formatting issue

**Review by srp33 — APPROVED** [review]

**Pull request merged**